text/0026-sassafras-consensus.md

```diff
@@ -0,0 +1,991 @@
+option allows for anonymous signatures using a key from a predefined set of
+enabled keys, known as the ring.
+
+#### 5.4.1. Plain VRF Signature
+
+This section describes the signature process for `VrfSignatureData` using the
+plain Bandersnatch signature flavor.
+
+```rust
+    PlainSignature ::= OCTET_STRING;
+
+    VrfSignature ::= SEQUENCE {
+        signature: PlainSignature,
+        outputs: SEQUENCE-OF VrfOutput
+    }
+```
+
+- `signature`: represents the actual signature (opaque).
+- `outputs`: a sequence of `VrfOutput`s corresponding to the `VrfInput`s values.
+
+Helper function to create a `VrfSignature` from `VrfSignatureData`:
+
+```rust
+    fn plain_vrf_sign(
+        secret: BandernatchSecretKey,
+        signature_data: VrfSignatureData
+    ) -> VrfSignature
+```
+
+Helper function for validating the signature and returning a Boolean value
+indicating the validity of the signature (`True` if it's valid):
+
+```rust
+    fn plain_vrf_verify(
+        public: BandersnatchPublicKey,
+        signature: VrfSignature
+    ) -> Boolean;
+```
+
+In this document, the types `BandersnatchSecretKey`, `BandersnatchPublicKey`
+and `PlainSignature` are intentionally left undefined as are not relevant. Their
+definitions can be found in the `bandersnatch_vrfs` reference implementation.
+
+#### 5.3.2. Ring VRF Signature
+
+This section deals with the signature process for `VrfSignatureData` using the
+Bandersnatch Ring signature flavor.
+
+```rust
+    RingSignature ::= OCTET_STRING;
+
+    RingVrfSignature ::= SEQUENCE {
+        signature: RingSignature,
+        ring_proof: RingProof,
+        outputs: SEQUENCE_OF VrfOutput
+    }
+```
+
+- `signature`: represents the actual signature (opaque).
+- `ring_proof`: denotes the proof of ring membership.
+- `outputs`: sequence of `VrfOutput` objects corresponding to the `VrfInput` values.
+
+Helper function to create a `RingVrfSignature` from `VrfSignatureData`:
+
+```rust
+    fn ring_vrf_sign(
+        secret: BandersnatchSecretKey,
+        signature_data: VrfSignatureData,
+        prover: RingProver
+    ) -> RingVrfSignature;
+```
+
+Helper function for validating the signature and returning a Boolean value
+indicating the validity of the signature (`True` if it's valid).
+It's important to note that this function does not require the signer's public key.
+
+```rust
+    fn ring_vrf_verify(
+        signature: RingVrfSignature,
+        verifier: RingVerifier
+    ) -> Boolean;
+```
+
+In this document, the types `BandersnatchSecretKey`, `BandersnatchPublicKey`,
+`RingSignature`, `RingProof`, `RingProver` and `RingVerifier` are intentionally
+left undefined as are not relevant. Their definitions can be found in the
+`bandersnatch_vrfs` reference implementation.
+
+
+## 6. Sassafras Protocol
+
+### 6.1 Epoch's First Block
+
+The first block produced for an epoch `N` is required to include the descriptor
+for the next epoch `N+1`.
```

Comment:
What about block 0? Does it belong to an epoch? Does it contain a descriptor?

Reply:
Also, this is more copying this information in state than including "in the block", but whatever.